src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -538,16 +583,145 @@ class ExtentPlacementManager {
+      }
+    }
+
+    struct details {
+      template<std::size_t, typename T>
+      using T_ = T;
+
+      template<typename T, std::size_t... I>
+      static constexpr auto create_tuple_impl(std::index_sequence<I...>) {
+        return std::tuple<T_<I, T>...>{};
+      }
+
+      template<typename T, std::size_t N>
+      static constexpr auto create_tuple() {
+        return create_tuple_impl<T>(std::make_index_sequence<N>{});
+      }
+
+      template<typename Tuple, typename Func>
+      static auto iterate_tuple_with_result(Tuple&& tuple, Func&& func) {
+        using TupleType = std::remove_reference_t<Tuple>;
+        using FirstTupleType = std::add_lvalue_reference_t<
+          std::tuple_element_t<0, TupleType>>;
+        using FuncReturnType = std::invoke_result_t<
+          Func, FirstTupleType, std::size_t>;
+        constexpr std::size_t tuple_size = std::tuple_size_v<TupleType>;
+
+        return []<std::size_t... I>(Tuple&& tuple, Func&& func,
+                                    std::index_sequence<I...>) {
+          auto res = create_tuple<FuncReturnType, tuple_size>();
+          ([&func, &tuple, &res]() {
+            std::get<I>(res) = func(std::get<I>(std::forward<Tuple>(tuple)), I);
+          }(), ...);
+          return res;
+        }(std::forward<Tuple>(tuple), std::forward<Func>(func),
+          std::make_index_sequence<tuple_size>{});
+      }
+
+      template<typename Tuple, typename Func>
+      static void iterate_tuple_without_result(Tuple&& tuple, Func&& func) {
+        using TupleType = std::remove_reference_t<Tuple>;
+        constexpr std::size_t tuple_size = std::tuple_size_v<TupleType>;
+
+        return []<std::size_t... I>(Tuple&& tuple, Func&& func,
+                                    std::index_sequence<I...>) {
+          (func(std::get<I>(std::forward<Tuple>(tuple)), I), ...);
+        }(std::forward<Tuple>(tuple), std::forward<Func>(func),
+          std::make_index_sequence<tuple_size>{});
+      }
+    };
+
+    template<typename... T>
+    auto try_reserve_impl(T&&... t) {
+      auto handler_tuple = std::make_tuple(create_reserve_handler(t)...);
+      std::set<std::size_t> failed_index;
+      auto res = details::iterate_tuple_with_result(
+        handler_tuple,
+        [&failed_index](auto &h, std::size_t idx) {
+          auto success = h->try_reserve();
+          if (!success) {
+            failed_index.insert(idx);
+          }
+          return success;
+        }
+      );
+
+      if (!failed_index.empty()) {
+        details::iterate_tuple_without_result(
+          handler_tuple,
+          [&failed_index](auto &h, std::size_t idx) {
+            if (!failed_index.contains(idx)) {
+              h->abort_reserve();
+            }
+          }
+        );
+      }
+
+      return res;
+    }
+
+    struct reserve_io_result_t {
       bool reserve_inline_success = true;
-      bool reserve_ool_success = true;
+      bool reserve_main_ool_success = true;
```

Comment:
The original logic was reserving both inline_usage and main_ool_usage.

This is why in https://github.com/ceph/ceph/pull/47974#discussion_r1061292389, the name is `reserve_cleaner_result_t::reserve_main_success`, not `reserve_main_ool_success`.